README.md
# CHIP-8 Emulator

![CHIP-8 Assembler + Emulator](https://raw.github.com/massung/CHIP-8/master/data/screencap.gif)

CHIP-8 is an assembler, debugger, and emulator for the [COSMAC ELF][2] [CHIP-8][3] interpreter and its derivative: the Super CHIP-8, which ran on HP-48 calculators. Everything is emulated as well as possible: the video display refreshes at 60 Hz and sound is emulated as well.

From the screen capture above you can see the disassembled program, register values, and a log which is used to show breakpoint information, memory dumps, and more.

CHIP-8 is written in [Go](https://golang.org/) and uses [SDL](https://www.libsdl.org/) for its rendering, input handling, and audio. It should easily run on Windows, OS X, and Linux. It - and the source code - are under the [ZLIB license](https://en.wikipedia.org/wiki/Zlib_License).

## What's New?

Getting the latest version? I recommend looking at the [CHANGELOG](CHANGELOG.md) file.

## Downloading

You can download the latest pre-built release for your platform [here](https://github.com/massung/CHIP-8/releases).

## Building

If you'd like to build yourself, then you'll need the [go-sdl2](https://godoc.org/github.com/veandco/go-sdl2/sdl) package (which requires [cgo](https://golang.org/cmd/cgo/) to build). But, once you have that it should be as simple as running:

```
$ go build -ldflags "-H windowsgui"
```

If you'd like a nice application icon for Windows, then grab [akavel/rsrc](https://github.com/akavel/rsrc) and build the `syso` file for your architecture and then build:

```
$ rsrc -arch [386|amd64] -ico chip-8.ico -o chip-8.syso
Icon  chip-8.ico  ID:  4

$ go build -ldflags "-H windowsgui"
```

That's it. You should have a `chip-8` executable ready to run.

## Usage

Simply launch the app and away you go!
 
If you get a panic on startup or the application fails to launch for an unknown reason, it's most likely due to one of the following:

* `SDL2.DLL` is missing (or not installed).
* `FONT.BMP` is missing.

Once launched simply drag a ROM or C8 source file into the app to load it. You can also press `H` at any time to see the list of key commands available to you. But here's a quick breakdown:

| Emulation         | Description
|:------------------|:-----------------
| `Back`            | Reset the emulator
| `Ctrl`+`Back`     | Reset and break
| `[`               | Decrease emulation speed
| `]`               | Increase emulation speed
| `F2`              | Reload ROM or C8 assembler file
| `F3`              | Open ROM or C8 assembler file
| `F4`              | Save ROM

| Debugging         | Description
|:------------------|:-----------------
| `F5`              | Pause/break emulation
| `F6`              | Step over
| `F7`              | Step into
| `SHIFT`+`F7`      | Step out
| `F8`              | Dump memory at `I` register
| `F9`              | Toggle breakpoint

_Note: You can launch the emulator with `-eti`. This will tell the emulator to assemble and load ROMs in a mode that supports the ETI-660. This flag should be rarely used. The ETI-660 loads CHIP-8 programs starting at address 0x600 instead of 0x200. Use this if you intend to assemble and run a ROM on actual ETI-660 hardware or if you have a ROM assembled for the ETI (good luck finding one!)._

### Virtual Key Mapping

The original CHIP-8 had 16 virtual keys had the layout on the left, which has been mapped (by default) to the keyboard layout on the right:

```
 1 2 3 C                                   1 2 3 4
 4 5 6 D    This is emulated with these    Q W E R
 7 8 9 E    keyboard keys -->              A S D F
 A 0 B F                                   Z X C V
```

## The Assembler

While playing the games that exist for the CHIP-8 might be fun for a while, the real fun is in creating your own games and seeing just how creative you can be with such a limited machine!

Just about every assembler for the CHIP-8 is different, and this one is, too. It's designed with a few niceties in mind. So, bear this in mind and take a few minutes to peruse this documentation.

Heavily documented, example programs can be found in [games/sources/](https://github.com/massung/CHIP-8/tree/master/games/sources).

### Syntax

Each line of assembly uses the following syntax:

```
label instruction arg0, arg1, ... ; comment
```

A label **must** appear at the very beginning of the line, and there **must** be at least a single whitespace character before the instruction or directive of a line (i.e. an instruction cannot appear at the beginning of a line). There are plenty of examples in `games/sources` to learn from.

### Registers

The CHIP-8 has 16, 8-bit virtual registers: `V0`, `V1`, `V2`, `V3`, `V4`, `V5`, `V6`, `V7`, `V8`, `V9`, `VA`, `VB`, `VC`, `VD`, `VE`, and `VF`. All of these are considered general purpose registers except for `VF` which is used for carry, borrow, shift, overflow, and collision detection.
 
There is a single, 16-bit address register: `I`, which is used for reading from - and writing to - memory.

Last, there are two, 8-bit, timer registers (`DT` for delays and `ST` for sounds) that continuously count down at 60 Hz. The delay timer is good for time limiting your game or waiting brief periods of time. While the sound timer is non-zero a tone will be emitted.

Finally, the Super CHIP-8, which was used on the HP-48 calculators, contained 8, 8-bit, user-flag registers: `R0`-`R7`. These cannot be directly used, but registers `V0`-`V7` can be saved to - and loaded from - them. This can be quite handy at times. See the `LD R, VX` and `LD VX, R` instructions below.

### Literals


CHANGELOG.md
# Changes

Want to know what's different between versions? Look no further...

## Version 1.3

___Fixes___

* Updated the code to match the latest go-sdl2 release.
* Touched up the comments to change `///` to `//`.

## Version 1.2

___Fixes___

* Window creation flags error with latest SDL. Thanks to @renatorabelo for finding it!

## Version 1.1

___Fixes___

* Fixed errors in README.
* Fixed scanning of string tokens.

___Additions___

* Added CHIP-8E instruction set.
* Added `SUPER` directive, required to use CHIP-48 instructions.
* Added `EXTENDED` directive, required to use CHIP-8E instructions.
* Added `ASCII` directive, for use with CHIP-8E `LD A,VX` instruction.
* Added use of back-quote (`) for text strings.
* Added Step Out functionality (`SHIFT`+`F7`) to debugger.

___Breaking Changes___

* Use of the `SUPER` directive is now required before using CHIP-48 instructions.
* Changed `LD B,VX` instruction to `BCD VX` to match CHIP-8E `BCD VX,VY`.

## Version 1.0

Initial release.


chip8/asm.go
/* Copyright (c) 2017 Jeffrey Massung
 *
 * This software is provided 'as-is', without any express or implied
 * warranty.  In no event will the authors be held liable for any damages
 * arising from the use of this software.
 *
 * Permission is granted to anyone to use this software for any purpose,
 * including commercial applications, and to alter it and redistribute it
 * freely, subject to the following restrictions:
 *
 * 1. The origin of this software must not be misrepresented; you must not
 *    claim that you wrote the original software. If you use this software
 *    in a product, an acknowledgment in the product documentation would be
 *    appreciated but is not required.
 *
 * 2. Altered source versions must be plainly marked as such, and must not be
 *    misrepresented as being the original software.
 *
 * 3. This notice may not be removed or altered from any source distribution.
 */

package chip8

import (
	"bufio"
	"bytes"
	"fmt"
	"strings"
)

// Assembly is a completely assembled source file.
type Assembly struct {
	// ROM is the final, assembled bytes to load.
	ROM []byte

	// Breakpoints is a list of addresses.
	Breakpoints []Breakpoint

	// Label mapping.
	Labels map[string]token

	// Addresses with unresolved labels.
	Unresolved map[int]string

	// Base address the ROM begins at (0x200 or 0x600 for ETI).
	Base int

	// Super is true if using additional super CHIP-8 instructions.
	Super bool

	// Extended is true if using additional CHIP-8E instructions.
	Extended bool
}

var (
	// AsciiTable is the 6-bit ASCII table for CHIP-8E.
	AsciiTable = `@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\]^_ !"#$%&'()*+,-./0123456789:;<=>?`
)

// Assemble an input CHIP-8 source code file.
func Assemble(program []byte, eti bool) (out *Assembly, err error) {
	var line int

	// base address for program
	base := 0x200

	// ETI-660 binaries are loaded to 0x600
	if eti {
		base = 0x600
	}

	// create an empty, return assembly
	out = &Assembly{
		ROM:         make([]byte, base, 0x1000),
		Breakpoints: make([]Breakpoint, 0, 10),
		Labels:      make(map[string]token),
		Unresolved:  make(map[int]string),
		Base:        base,
	}

	// no error
	err = nil

	// handle panics during assembly
	defer func() {
		if r := recover(); r != nil {
			if line > 0 {
				err = fmt.Errorf("line %d - %s", line, r)
			} else {
				err = fmt.Errorf("%s", r)
			}

			// return a dummy ROM
			out = &Assembly{ROM: Dummy}
		}
	}()

	// create simple line scanner over the file
	reader := bytes.NewReader(bytes.ToUpper(program))
	scanner := bufio.NewScanner(reader)

	// parse and assemble
	for line = 1; scanner.Scan(); line++ {
		out.assemble(&tokenScanner{bytes: scanner.Bytes()})
	}

	// resolve all label addresses
	for address, label := range out.Unresolved {
		if t, ok := out.Labels[label]; ok {
			if t.typ != TOKEN_LIT {
				panic("label does not resolve to address!")
			}

			msb := byte(t.val.(int) >> 8)
			lsb := byte(t.val.(int) & 0xFF)

			// NOTE: This "just works" because all labels are guaranteed to be
			//       addressed within 12-bits. There are only a handful of
			//       instructions that take an immediate Address:
			//         SYS    NNN
			//         CALL   NNN
			//         JP     NNN
			//         JP     V0, NNN
			//         LD     I, NNN
			//       The only other use case is the WORD instruction to write
			//       16-bit values, and since the unresolved label defaulted
			//       to 0x0200, overwriting it works just fine.
			out.ROM[address] = msb | (out.ROM[address] & 0xF0)
			out.ROM[address+1] = lsb

			// delete the unresolved Address
			delete(out.Unresolved, address)
		}
	}

	// clear the line number as we're done assembling
	line = 0

	// if there are any unresolved addresses, panic
	for _, label := range out.Unresolved {
		panic(fmt.Errorf("unresolved label: %s", label))
	}

	// drop the first 512 bytes from the rom
	out.ROM = out.ROM[base:]

	// done
	return
}

// Compile a single line into the assembly.
func (a *Assembly) assemble(s *tokenScanner) {
	t := s.scanToken()

	// assign labels
	if t.typ == TOKEN_LABEL {
		t = a.assembleLabel(t.val.(string), s)
	}

	// continue assembling
	switch {
	case t.typ == TOKEN_INSTRUCTION:
		a.assembleInstruction(t.val.(string), s)
	case t.typ == TOKEN_SUPER:
		a.assembleSuper(s)
	case t.typ == TOKEN_EXTENDED:
		a.assembleExtended(s)
	case t.typ == TOKEN_BREAK:
		a.assembleBreakpoint(s, false)
	case t.typ == TOKEN_ASSERT:
		a.assembleBreakpoint(s, true)
	case t.typ != TOKEN_END:
		panic("unexpected token")
	}

}

// Scan for a label and add it to the assembly.
func (a *Assembly) assembleLabel(label string, s *tokenScanner) token {
	if _, exists := a.Labels[label]; exists {
		panic("duplicate label")
	}

	// by default, the label is assigned the current address
	a.Labels[label] = token{typ: TOKEN_LIT, val: len(a.ROM)}

	// scan the next token
	t := s.scanToken()

	// if EQU or VAR, reassign the label
	if t.typ == TOKEN_EQU || t.typ == TOKEN_VAR {
		v := s.scanToken()

		// equ requires a literal, and var requires a v-register
		if (t.typ == TOKEN_EQU && v.typ == TOKEN_LIT) || (t.typ == TOKEN_VAR && v.typ == TOKEN_V) {
			a.Labels[label] = v

			// should be the final token
			if t = s.scanToken(); t.typ == TOKEN_END {
				return t
			}
		}

		panic("illegal label assignment")
	}

	return t
}

// Create a new breakpoint at the current Address.
func (a *Assembly) assembleBreakpoint(s *tokenScanner, conditional bool) {
	reason := s.scanToEnd().val.(string)

	// create the breakpoint
	a.Breakpoints = append(a.Breakpoints, Breakpoint{
		Address:     len(a.ROM),
		Conditional: conditional,
		Reason:      reason,
	})
}

// Allow the assembler to assemble super, SCHIP-8 instructions.
func (a *Assembly) assembleSuper(s *tokenScanner) {
	if s.scanToken().typ != TOKEN_END {
		panic("unexpected token")
	}

	if len(a.ROM) > a.Base {
		panic("super must come before instructions")
	}

	// enter super instructions mode
	a.Super = true
}

// Allow the assembler to assemble extended, CHIP-8E instructions.
func (a *Assembly) assembleExtended(s *tokenScanner) {
	if s.scanToken().typ != TOKEN_END {
		panic("unexpected token")
	}

	if len(a.ROM) > a.Base {
		panic("extended must come before instructions")
	}

	// enter extended instructions mode
	a.Extended = true
}

// Compile a single instruction into the assembly.
func (a *Assembly) assembleInstruction(i string, s *tokenScanner) {
	tokens := s.scanOperands()

	switch i {
	case "CLS":
		a.ROM = append(a.ROM, a.assembleCLS(tokens)...)
	case "RET":
		a.ROM = append(a.ROM, a.assembleRET(tokens)...)
	case "EXIT":
		a.ROM = append(a.ROM, a.assembleEXIT(tokens)...)
	case "LOW":
		a.ROM = append(a.ROM, a.assembleLOW(tokens)...)
	case "HIGH":
		a.ROM = append(a.ROM, a.assembleHIGH(tokens)...)
	case "SCU":
		a.ROM = append(a.ROM, a.assembleSCU(tokens)...)
	case "SCD":
		a.ROM = append(a.ROM, a.assembleSCD(tokens)...)
	case "SCR":
		a.ROM = append(a.ROM, a.assembleSCR(tokens)...)
	case "SCL":
		a.ROM = append(a.ROM, a.assembleSCL(tokens)...)
	case "SYS":
		a.ROM = append(a.ROM, a.assembleSYS(tokens)...)
	case "JP":
		a.ROM = append(a.ROM, a.assembleJP(tokens)...)
	case "CALL":
		a.ROM = append(a.ROM, a.assembleCALL(tokens)...)
	case "SE":
		a.ROM = append(a.ROM, a.assembleSE(tokens)...)
	case "SNE":
		a.ROM = append(a.ROM, a.assembleSNE(tokens)...)
	case "SGT":
		a.ROM = append(a.ROM, a.assembleSGT(tokens)...)
	case "SLT":
		a.ROM = append(a.ROM, a.assembleSLT(tokens)...)
	case "SKP":
		a.ROM = append(a.ROM, a.assembleSKP(tokens)...)
	case "SKNP":
		a.ROM = append(a.ROM, a.assembleSKNP(tokens)...)
	case "OR":
		a.ROM = append(a.ROM, a.assembleOR(tokens)...)
	case "AND":
		a.ROM = append(a.ROM, a.assembleAND(tokens)...)
	case "XOR":
		a.ROM = append(a.ROM, a.assembleXOR(tokens)...)
	case "SHR":
		a.ROM = append(a.ROM, a.assembleSHR(tokens)...)
	case "SHL":
		a.ROM = append(a.ROM, a.assembleSHL(tokens)...)
	case "ADD":
		a.ROM = append(a.ROM, a.assembleADD(tokens)...)
	case "SUB":
		a.ROM = append(a.ROM, a.assembleSUB(tokens)...)
	case "SUBN":
		a.ROM = append(a.ROM, a.assembleSUBN(tokens)...)
	case "MUL":
		a.ROM = append(a.ROM, a.assembleMUL(tokens)...)
	case "DIV":
		a.ROM = append(a.ROM, a.assembleDIV(tokens)...)
	case "BCD":
		a.ROM = append(a.ROM, a.assembleBCD(tokens)...)
	case "RND":
		a.ROM = append(a.ROM, a.assembleRND(tokens)...)
	case "DRW":
		a.ROM = append(a.ROM, a.assembleDRW(tokens)...)
	case "LD":
		a.ROM = append(a.ROM, a.assembleLD(tokens)...)
	case "ASCII":
		a.ROM = append(a.ROM, a.assembleASCII(tokens)...)
	case "BYTE":
		a.ROM = append(a.ROM, a.assembleBYTE(tokens)...)
	case "WORD":
		a.ROM = append(a.ROM, a.assembleWORD(tokens)...)
	case "ALIGN":
		a.ROM = append(a.ROM, a.assembleALIGN(tokens)...)
	case "PAD":
		a.ROM = append(a.ROM, a.assemblePAD(tokens)...)
	}
}

// Assemble a single operand, expanding label references.
func (a *Assembly) assembleOperand(t token) token {
	if t.typ == TOKEN_ID {
		label := t.val.(string)
		if v, exists := a.Labels[label]; exists {
			t = v
		} else {
			t = token{typ: TOKEN_LIT, val: 0x200}

			// add an unresolved address
			a.Unresolved[len(a.ROM)] = label
		}
	}

	return t
}

// Match the desired tokens with a list of tokens. Expand defines and labels.
func (a *Assembly) assembleOperands(tokens []token, m ...tokenType) ([]token, bool) {
	ops := make([]token, 0, 3)

	// the number of desired tokens should match
	if len(tokens) != len(m) {
		return nil, false
	}

	// expand and compare the token types
	for i, typ := range m {
		t := a.assembleOperand(tokens[i])

		// compare token types
		if t.typ != typ {
			return nil, false
		}

		// append the operand
		ops = append(ops, t)
	}

	return ops, true
}

// Assemble a CLS instruction.
func (a *Assembly) assembleCLS(tokens []token) []byte {
	if len(tokens) == 0 {
		return []byte{0x00, 0xE0}
	}

	panic("illegal instruction")
}

// Assemble a RET instruction.
func (a *Assembly) assembleRET(tokens []token) []byte {
	if len(tokens) == 0 {
		return []byte{0x00, 0xEE}
	}

	panic("illegal instruction")
}

// Assemble an EXIT instruction.
func (a *Assembly) assembleEXIT(tokens []token) []byte {
	if a.Super {
		if len(tokens) == 0 {
			return []byte{0x00, 0xFD}
		}
	}

	panic("illegal instruction")
}

// Assemble a LOW instruction.
func (a *Assembly) assembleLOW(tokens []token) []byte {
	if a.Super {
		if len(tokens) == 0 {
			return []byte{0x00, 0xFE}
		}
	}

	panic("illegal instruction")
}

// Assemble a HIGH instruction.
func (a *Assembly) assembleHIGH(tokens []token) []byte {
	if a.Super {
		if len(tokens) == 0 {
			return []byte{0x00, 0xFF}
		}
	}

	panic("illegal instruction")
}

// Assemble a SCU instruction.
func (a *Assembly) assembleSCU(tokens []token) []byte {
	if a.Super {
		if ops, ok := a.assembleOperands(tokens, TOKEN_LIT); ok {
			n := ops[0].val.(int)

			if n < 0x10 {
				return []byte{0x00, 0xB0 | byte(n)}
			}
		}
	}

	panic("illegal instruction")
}

// Assemble a SCD instruction.
func (a *Assembly) assembleSCD(tokens []token) []byte {
	if a.Super {
		if ops, ok := a.assembleOperands(tokens, TOKEN_LIT); ok {
			n := ops[0].val.(int)

			if n < 0x10 {
				return []byte{0x00, 0xC0 | byte(n)}
			}
		}
	}

	panic("illegal instruction")
}

// Assemble a SCR instruction.
func (a *Assembly) assembleSCR(tokens []token) []byte {
	if a.Super {
		if len(tokens) == 0 {
			return []byte{0x00, 0xFB}
		}
	}

	panic("illegal instruction")
}

// Assemble a SCL instruction.
func (a *Assembly) assembleSCL(tokens []token) []byte {
	if a.Super {
		if len(tokens) == 0 {
			return []byte{0x00, 0xFC}
		}
	}

	panic("illegal instruction")
}

// Assemble a SYS instruction.
func (a *Assembly) assembleSYS(tokens []token) []byte {
	if ops, ok := a.assembleOperands(tokens, TOKEN_LIT); ok {
		a := ops[0].val.(int)

		if a < 0x1000 {
			return []byte{byte(a >> 8 & 0xF), byte(a & 0xFF)}
		}
	}

	panic("illegal instruction")
}

// Assemble a JP instruction.
func (a *Assembly) assembleJP(tokens []token) []byte {
	if ops, ok := a.assembleOperands(tokens, TOKEN_LIT); ok {
		a := ops[0].val.(int)

		if a < 0x1000 {
			return []byte{0x10 | byte(a>>8&0xF), byte(a & 0xFF)}
		}
	}

	if ops, ok := a.assembleOperands(tokens, TOKEN_V, TOKEN_LIT); ok {
		v := ops[0].val.(int)
		a := ops[1].val.(int)

		if v == 0 && a < 0x1000 {
			return []byte{0xB0 | byte(a>>8&0xF), byte(a & 0xFF)}
		}
	}

	panic("illegal instruction")
}

// Assemble a CALL instruction.
func (a *Assembly) assembleCALL(tokens []token) []byte {
	if ops, ok := a.assembleOperands(tokens, TOKEN_LIT); ok {
		a := ops[0].val.(int)

		if a < 0x1000 {
			return []byte{0x20 | byte(a>>8&0xF), byte(a & 0xFF)}
		}
	}

	panic("illegal instruction")
}

// Assemble a SE instruction.
func (a *Assembly) assembleSE(tokens []token) []byte {
	if ops, ok := a.assembleOperands(tokens, TOKEN_V, TOKEN_LIT); ok {
		x := ops[0].val.(int)
		b := ops[1].val.(int)

		if b < 0x100 {
			return []byte{0x30 | byte(x), byte(b)}
		}
	}

	if ops, ok := a.assembleOperands(tokens, TOKEN_V, TOKEN_V); ok {
		x := ops[0].val.(int)
		y := ops[1].val.(int)

		return []byte{0x50 | byte(x), byte(y << 4)}
	}

	panic("illegal instruction")
}

// Assemble a SNE instruction.
func (a *Assembly) assembleSNE(tokens []token) []byte {
	if ops, ok := a.assembleOperands(tokens, TOKEN_V, TOKEN_LIT); ok {
		x := ops[0].val.(int)
		b := ops[1].val.(int)

		if b < 0x100 {
			return []byte{0x40 | byte(x), byte(b)}
		}
	}

	if ops, ok := a.assembleOperands(tokens, TOKEN_V, TOKEN_V); ok {
		x := ops[0].val.(int)
		y := ops[1].val.(int)

		return []byte{0x90 | byte(x), byte(y << 4)}
	}

	panic("illegal instruction")
}

// Assemble a SGT instruction.
func (a *Assembly) assembleSGT(tokens []token) []byte {
	if a.Extended {
		if ops, ok := a.assembleOperands(tokens, TOKEN_V, TOKEN_V); ok {
			x := ops[0].val.(int)
			y := ops[1].val.(int)

			return []byte{0x50 | byte(x), byte(y<<4) | 0x01}
		}
	}

	panic("illegal instruction")
}

// Assemble a SLT instruction.
func (a *Assembly) assembleSLT(tokens []token) []byte {
	if a.Extended {
		if ops, ok := a.assembleOperands(tokens, TOKEN_V, TOKEN_V); ok {
			x := ops[0].val.(int)
			y := ops[1].val.(int)

			return []byte{0x50 | byte(x), byte(y<<4) | 0x02}
		}
	}

	panic("illegal instruction")
}

// Assemble a SKP instruction.
func (a *Assembly) assembleSKP(tokens []token) []byte {
	if ops, ok := a.assembleOperands(tokens, TOKEN_V); ok {
		x := ops[0].val.(int)

		return []byte{0xE0 | byte(x), 0x9E}
	}

	panic("illegal instruction")
}

// Assemble a SKNP instruction.
func (a *Assembly) assembleSKNP(tokens []token) []byte {
	if ops, ok := a.assembleOperands(tokens, TOKEN_V); ok {
		x := ops[0].val.(int)

		return []byte{0xE0 | byte(x), 0xA1}
	}

	panic("illegal instruction")
}

// Assemble a OR instruction.
func (a *Assembly) assembleOR(tokens []token) []byte {
	if ops, ok := a.assembleOperands(tokens, TOKEN_V, TOKEN_V); ok {
		x := ops[0].val.(int)
		y := ops[1].val.(int)

		return []byte{0x80 | byte(x), byte(y<<4) | 0x01}
	}

	panic("illegal instruction")
}

// Assemble a AND instruction.
func (a *Assembly) assembleAND(tokens []token) []byte {
	if ops, ok := a.assembleOperands(tokens, TOKEN_V, TOKEN_V); ok {
		x := ops[0].val.(int)
		y := ops[1].val.(int)

		return []byte{0x80 | byte(x), byte(y<<4) | 0x02}
	}

	panic("illegal instruction")
}

// Assemble a XOR instruction.
func (a *Assembly) assembleXOR(tokens []token) []byte {
	if ops, ok := a.assembleOperands(tokens, TOKEN_V, TOKEN_V); ok {
		x := ops[0].val.(int)
		y := ops[1].val.(int)

		return []byte{0x80 | byte(x), byte(y<<4) | 0x03}
	}

	panic("illegal instruction")
}

// Assemble a SHR instruction.
func (a *Assembly) assembleSHR(tokens []token) []byte {
	if ops, ok := a.assembleOperands(tokens, TOKEN_V); ok {
		x := ops[0].val.(int)

		return []byte{0x80 | byte(x), byte(x<<4) | 0x06}
	}

	panic("illegal instruction")
}

// Assemble a SHL instruction.
func (a *Assembly) assembleSHL(tokens []token) []byte {
	if ops, ok := a.assembleOperands(tokens, TOKEN_V); ok {
		x := ops[0].val.(int)

		return []byte{0x80 | byte(x), byte(x<<4) | 0x0E}
	}

	panic("illegal instruction")
}

// Assemble a ADD instruction.
func (a *Assembly) assembleADD(tokens []token) []byte {
	if ops, ok := a.assembleOperands(tokens, TOKEN_V, TOKEN_LIT); ok {
		x := ops[0].val.(int)
		b := ops[1].val.(int)

		if b < 0x100 {
			return []byte{0x70 | byte(x), byte(b)}
		}
	}

	if ops, ok := a.assembleOperands(tokens, TOKEN_V, TOKEN_V); ok {
		x := ops[0].val.(int)
		y := ops[1].val.(int)

		return []byte{0x80 | byte(x), byte(y<<4) | 0x04}
	}

	if ops, ok := a.assembleOperands(tokens, TOKEN_I, TOKEN_V); ok {
		x := ops[1].val.(int)

		return []byte{0xF0 | byte(x), 0x1E}
	}

	panic("illegal instruction")
}

// Assemble a SUB instruction.
func (a *Assembly) assembleSUB(tokens []token) []byte {
	if ops, ok := a.assembleOperands(tokens, TOKEN_V, TOKEN_V); ok {
		x := ops[0].val.(int)
		y := ops[1].val.(int)

		return []byte{0x80 | byte(x), byte(y<<4) | 0x05}
	}

	panic("illegal instruction")
}

// Assemble a SUBN instruction.
func (a *Assembly) assembleSUBN(tokens []token) []byte {
	if ops, ok := a.assembleOperands(tokens, TOKEN_V, TOKEN_V); ok {
		x := ops[0].val.(int)
		y := ops[1].val.(int)

		return []byte{0x80 | byte(x), byte(y<<4) | 0x07}
	}

	panic("illegal instruction")
}

// Assemble a MUL instruction.
func (a *Assembly) assembleMUL(tokens []token) []byte {
	if a.Extended {
		if ops, ok := a.assembleOperands(tokens, TOKEN_V, TOKEN_V); ok {
			x := ops[0].val.(int)
			y := ops[1].val.(int)

			return []byte{0x90 | byte(x), byte(y<<4) | 0x01}
		}
	}

	panic("illegal instruction")
}

// Assemble a DIV instruction.
func (a *Assembly) assembleDIV(tokens []token) []byte {
	if a.Extended {
		if ops, ok := a.assembleOperands(tokens, TOKEN_V, TOKEN_V); ok {
			x := ops[0].val.(int)
			y := ops[1].val.(int)

			return []byte{0x90 | byte(x), byte(y<<4) | 0x02}
		}
	}

	panic("illegal instruction")
}

// Assemble a BCD instruction.
func (a *Assembly) assembleBCD(tokens []token) []byte {
	if ops, ok := a.assembleOperands(tokens, TOKEN_V); ok {
		x := ops[0].val.(int)

		return []byte{0xF0 | byte(x), 0x33}
	}

	if a.Extended {
		if ops, ok := a.assembleOperands(tokens, TOKEN_V, TOKEN_V); ok {
			x := ops[0].val.(int)
			y := ops[1].val.(int)

			return []byte{0x90 | byte(x), byte(y<<4) | 0x03}
		}
	}

	panic("illegal instruction")
}

// Assemble a RND instruction.
func (a *Assembly) assembleRND(tokens []token) []byte {
	if ops, ok := a.assembleOperands(tokens, TOKEN_V, TOKEN_LIT); ok {
		x := ops[0].val.(int)
		b := ops[1].val.(int)

		if b < 0x100 {
			return []byte{0xC0 | byte(x), byte(b)}
		}
	}

	panic("illegal instruction")
}

// Assemble a DRW instruction.
func (a *Assembly) assembleDRW(tokens []token) []byte {
	if ops, ok := a.assembleOperands(tokens, TOKEN_V, TOKEN_V, TOKEN_LIT); ok {
		x := ops[0].val.(int)
		y := ops[1].val.(int)
		n := ops[2].val.(int)

		if n < 0x10 {
			return []byte{0xD0 | byte(x), byte(y<<4) | byte(n)}
		}
	}

	panic("illegal instruction")
}

// Assemble a LD instruction.
func (a *Assembly) assembleLD(tokens []token) []byte {
	if ops, ok := a.assembleOperands(tokens, TOKEN_V, TOKEN_LIT); ok {
		x := ops[0].val.(int)
		b := ops[1].val.(int)

		if b < 0x100 {
			return []byte{0x60 | byte(x), byte(b)}
		}
	}

	if ops, ok := a.assembleOperands(tokens, TOKEN_V, TOKEN_V); ok {
		x := ops[0].val.(int)
		y := ops[1].val.(int)

		return []byte{0x80 | byte(x), byte(y << 4)}
	}

	if ops, ok := a.assembleOperands(tokens, TOKEN_I, TOKEN_LIT); ok {
		a := ops[1].val.(int)

		if a < 0x1000 {
			return []byte{0xA0 | byte(a>>8&0xF), byte(a & 0xFF)}
		}
	}

	if ops, ok := a.assembleOperands(tokens, TOKEN_V, TOKEN_DT); ok {
		x := ops[0].val.(int)

		return []byte{0xF0 | byte(x), 0x07}
	}

	if ops, ok := a.assembleOperands(tokens, TOKEN_V, TOKEN_K); ok {
		x := ops[0].val.(int)

		return []byte{0xF0 | byte(x), 0x0A}
	}

	if ops, ok := a.assembleOperands(tokens, TOKEN_DT, TOKEN_V); ok {
		x := ops[1].val.(int)

		return []byte{0xF0 | byte(x), 0x15}
	}

	if ops, ok := a.assembleOperands(tokens, TOKEN_ST, TOKEN_V); ok {
		x := ops[1].val.(int)

		return []byte{0xF0 | byte(x), 0x18}
	}

	if ops, ok := a.assembleOperands(tokens, TOKEN_F, TOKEN_V); ok {
		x := ops[1].val.(int)

		return []byte{0xF0 | byte(x), 0x29}
	}

	if ops, ok := a.assembleOperands(tokens, TOKEN_EFFECTIVE_ADDRESS, TOKEN_V); ok {
		x := ops[1].val.(int)

		return []byte{0xF0 | byte(x), 0x55}
	}

	if ops, ok := a.assembleOperands(tokens, TOKEN_V, TOKEN_EFFECTIVE_ADDRESS); ok {
		x := ops[0].val.(int)

		return []byte{0xF0 | byte(x), 0x65}
	}

	if a.Super {
		if ops, ok := a.assembleOperands(tokens, TOKEN_HF, TOKEN_V); ok {
			x := ops[1].val.(int)

			return []byte{0xF0 | byte(x), 0x30}
		}

		if ops, ok := a.assembleOperands(tokens, TOKEN_R, TOKEN_V); ok {
			x := ops[1].val.(int)

			if x < 8 {
				return []byte{0xF0 | byte(x), 0x75}
			}
		}

		if ops, ok := a.assembleOperands(tokens, TOKEN_V, TOKEN_R); ok {
			x := ops[0].val.(int)

			if x < 8 {
				return []byte{0xF0 | byte(x), 0x85}
			}
		}
	}

	if a.Extended {
		if ops, ok := a.assembleOperands(tokens, TOKEN_ASCII, TOKEN_V); ok {
			x := ops[1].val.(int)

			return []byte{0xF0 | byte(x), 0x94}
		}
	}

	panic("illegal instruction")
}

// Assemble an ASCII instruction.
func (a *Assembly) assembleASCII(tokens []token) []byte {
	var b []byte

	if !a.Extended {
		panic("illegal directive")
	}

	// loop over all string tokens and assemble them as 6-bit ascii
	for _, t := range tokens {
		op := a.assembleOperand(t)

		if op.typ != TOKEN_TEXT {
			panic("expected ascii string")
		}

		// loop over each byte in the string, write the ascii table value
		for _, c := range op.val.(string) {
			if i := strings.IndexRune(AsciiTable, c); i < 0 {
				panic("invalid CHIP-8E ascii character")
			} else {
				b = append(b, byte(i))
			}
		}
	}

	return b
}

// Assemble a BYTE instruction.
func (a *Assembly) assembleBYTE(tokens []token) []byte {
	b := make([]byte, 0)

	for _, t := range tokens {
		op := a.assembleOperand(t)

		switch op.typ {
		case TOKEN_LIT:
			if op.val.(int) > 0xFF {
				panic("invalid byte")
			}

			b = append(b, byte(t.val.(int)))
		case TOKEN_TEXT:
			b = append(b, op.val.(string)...)
		}

	}

	return b
}

// Assemble a WORD instruction.
func (a *Assembly) assembleWORD(tokens []token) []byte {
	b := make([]byte, 0)

	for _, t := range tokens {
		op := a.assembleOperand(t)

		if op.typ != TOKEN_LIT || op.val.(int) > 0xFFFF {
			panic("invalid word")
		}

		msb := op.val.(int) >> 8 & 0xFF
		lsb := op.val.(int) & 0xFF

		// store msb first
		b = append(b, byte(msb), byte(lsb))
	}

	return b
}

// Assemble an ALIGN instruction.
func (a *Assembly) assembleALIGN(tokens []token) []byte {
	if ops, ok := a.assembleOperands(tokens, TOKEN_LIT); ok {
		n := ops[0].val.(int)

		if n&(n-1) == 0 {
			offset := len(a.ROM) & (n - 1)
			pad := n - offset

			// reserve pad bytes to meet alignment
			return make([]byte, pad)
		}
	}

	panic("illegal alignment")
}

// Assemble an PAD instruction.
func (a *Assembly) assemblePAD(tokens []token) []byte {
	if ops, ok := a.assembleOperands(tokens, TOKEN_LIT); ok {
		n := ops[0].val.(int)

		if n < 0x1000-len(a.ROM) {
			return make([]byte, n)
		}
	}

	panic("illegal size")
}
